src/combinatorics/antichain.lean

```diff
@@ -0,0 +1,80 @@
+/-
+Copyright (c) 2020 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Bhavik Mehta, Alena Gusakov
+-/
+import data.finset
+import data.fintype.basic
+
+/-!
+# Antichains
+
+Basic definitions for finite sets which are useful for combinatorics.
+We define antichains, and a proposition asserting that a set is a set of r-sets.
+
+## Main definitions
+
+* `antichain` is a family of sets where no set is a subset of another.
+* `all_sized` is a proposition that
+
+-/
+
+open finset
+
+variable {α : Type*}
+variable {r : ℕ}
+
+/--
+A family of sets is an antichain if no set is a subset of another. For example,
+{{1}, {4,6,7}, {2,4,5,6}} is an antichain.
+-/
+def antichain (𝒜 : finset (finset α)) : Prop := ∀ A ∈ 𝒜, ∀ B ∈ 𝒜, A ≠ B → ¬(A ⊆ B)
+
+/-- `all_sized 𝒜 r` states that every set in 𝒜 has size r. -/
+@[reducible]
+def all_sized (𝒜 : finset (finset α)) (r : ℕ) : Prop := ∀ A ∈ 𝒜, card A = r
+
+lemma antichain_of_all_sized {𝒜 : finset (finset α)} {r : ℕ} (a : all_sized 𝒜 r) :
+  antichain 𝒜 :=
+begin
+  intros A h1 B h2 h3,
+  have h4 : card A = card B,
+  { rw a A h1,
+    rw a B h2},
+  contrapose! h3,
+  convert eq_of_subset_of_card_le h3 _,
+  rw h4,
+end
+
+/--
+All sets in the union have size `r` iff both sets individually have this
+property.
+-/
+lemma union_layer [decidable_eq α] {A B : finset (finset α)} :
+  all_sized A r ∧ all_sized B r ↔ all_sized (A ∪ B) r :=
```

Comment:
This looks to me like what's missing is `finset.forall_mem_union`, to match the `list.forall_mem_union` that already exists - perhaps as:
```
lemma finset.forall_mem_union [decidable_eq α] {A B : finset α} {p : α → Prop} :
  (∀ a ∈ A, p a) ∧ (∀ b ∈ B, p b) ↔ ∀ ab ∈ (A ∪ B), p ab :=
⟨λ h ab hab, (mem_union.mp hab).elim (h.1 _) (h.2 _),
 λ h, ⟨λ a, h a ∘ mem_union_left _, λ b, h b ∘ mem_union_right _⟩⟩
```